samples/fpgaperf_counter/fpgaperf_counter.c

```diff
@@ -0,0 +1,639 @@
+			if (ioctl(g_fpga_perf->perf_events[loop].fd, PERF_EVENT_IOC_ID,
+						&g_fpga_perf->perf_events[loop].id) == -1) {
+				OPAE_ERR("PERF_EVENT_IOC_ID ioctl failed: %s",
+						strerror(errno));
+				ret = FPGA_EXCEPTION;
+				goto out;
+			}
+		}
+	}
+	if (ioctl(g_fpga_perf->perf_events[0].fd, PERF_EVENT_IOC_RESET,
+				PERF_IOC_FLAG_GROUP) == -1) {
+		OPAE_ERR("PERF_EVENT_IOC_RESET ioctl failed: %s", strerror(errno));
+		ret = FPGA_EXCEPTION;
+		goto out;
+	}
+out:
+	udev_unref(udev);
+	udev_device_unref(dev);
+	opae_mutex_unlock(res, &fpga_perf_lock);
+	return ret;
+err:
+	udev_unref(udev);
+	udev_device_unref(dev);
+	return FPGA_EXCEPTION;
+}
+
+/* get fpga sbdf from token */
+fpga_result get_Fpga_Sbdf(fpga_token token,
+		uint16_t *segment,
+		uint8_t *bus,
+		uint8_t *device,
+		uint8_t *function)
+{
+	fpga_result res = FPGA_OK;
+	fpga_properties props = NULL;
+
+	if (!segment || !bus ||
+		!device || !function) {
+		OPAE_ERR("Invalid input parameters");
+		return FPGA_INVALID_PARAM;
+	}
+	res = fpgaGetProperties(token, &props);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get properties");
+		return res;
+	}
+	res = fpgaPropertiesGetBus(props, bus);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get bus");
+		return res;
+	}
+	res = fpgaPropertiesGetSegment(props, segment);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Segment");
+		return res;
+	}
+	res = fpgaPropertiesGetDevice(props, device);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Device");
+		return res;
+	}
+	res = fpgaPropertiesGetFunction(props, function);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get Function");
+		return res;
+	}
+
+	return res;
+}
+
+fpga_result fpgaPerfCounterEnum(fpga_token token)
+{
+	fpga_result ret				= FPGA_OK;
+	char sysfs_path[DFL_PERF_STR_MAX]	= { 0 };
+	char sysfs_perf[DFL_PERF_STR_MAX]	= { 0 };
+	int gres 				= 0;
+	int res 				= 0;
+	uint32_t fpga_id 			= -1;
+	char *endptr 				= NULL;
+	glob_t pglob;
+	uint8_t bus				= (uint8_t)-1;
+	uint16_t segment			= (uint16_t)-1;
+	uint8_t device				= (uint8_t)-1;
+	uint8_t function			= (uint8_t)-1;
+
+	res = get_Fpga_Sbdf(token, &segment, &bus, &device, &function);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to get sbdf ");
+		return res;
+	}
+	/* when we bind with new device id we will get updated function value */
+	/* not able to read the sysfs path using that */
+	if(function)
+		function = 0;
+	if (snprintf(sysfs_path, sizeof(sysfs_path),
+		DFL_PERF_FME,
+		segment, bus, device, function) < 0) {
+		OPAE_ERR("snprintf buffer overflow");
+		return FPGA_EXCEPTION;
+	}
+	gres = glob(sysfs_path, GLOB_NOSORT, NULL, &pglob);
+	if (gres) {
+		OPAE_ERR("Failed pattern match %s: %s", sysfs_path, strerror(errno));
+		globfree(&pglob);
+		return FPGA_NOT_FOUND;
+	}
+	if (pglob.gl_pathc == 1) {
+		char *ptr = strstr(pglob.gl_pathv[0], "fme");
+		if (!ptr) {
+			ret = FPGA_INVALID_PARAM;
+			goto out;
+		}
+		errno = 0;
+		fpga_id = strtoul(ptr + 3, &endptr, 16);
```

Comment:
Is this the number at the end? I'm not sure it's in hex.

Reply:
changed to decimal value